File: webapp-net/Search/Controllers/SearchController.cs
using System;
using System.Collections.Specialized;
using System.Linq;
using Sdl.Web.Common.Logging;
using Sdl.Web.Common.Models;
using Sdl.Web.Modules.Search.Models;
using Sdl.Web.Modules.Search.Providers;
using Sdl.Web.Mvc.Configuration;
using Sdl.Web.Mvc.Controllers;

namespace Sdl.Web.Modules.Search.Controllers
{
    public class SearchController : EntityController
    {
        protected ISearchProvider SearchProvider { get; set; }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="searchProvider">The Search Provider to use. Set using Dependency Injection (ISearchProvider interface must be defined in Unity.config)</param>
        public SearchController(ISearchProvider searchProvider)
        {
            SearchProvider = searchProvider;
        }

        /// <summary>
        /// Enrich the SearchQuery View Model with request querystring parameters and populate the results using a configured Search Provider.
        /// </summary>
        protected override ViewModel EnrichModel(ViewModel model)
        {
            using (new Tracer(model))
            {
                base.EnrichModel(model);

                SearchQuery searchQuery = model as SearchQuery;
                if (searchQuery == null || !searchQuery.GetType().IsGenericType)
                {
                    throw new DxaSearchException($"Unexpected View Model: '{model}'. Expecting type SearchQuery<T>.");
                }

                NameValueCollection queryString = Request.QueryString;
                // Map standard query string parameters
                searchQuery.QueryText = queryString["q"];
                searchQuery.Start = queryString.AllKeys.Contains("start") ? Convert.ToInt32(queryString["start"]) : 1;
                // To allow the Search Provider to use additional query string parameters:
                searchQuery.QueryStringParameters = queryString;

                Type searchItemType = searchQuery.GetType().GetGenericArguments()[0];
                SearchProvider.ExecuteQuery(searchQuery, searchItemType, WebRequestContext.Localization);

                return searchQuery;
            }
        }
    }
}

File: webapp-java/dxa-module-search/src/main/java/com/sdl/dxa/modules/search/controller/SearchController.java
package com.sdl.dxa.modules.search.controller;

import com.sdl.dxa.modules.search.DxaSearchException;
import com.sdl.dxa.modules.search.model.SearchQuery;
import com.sdl.dxa.modules.search.provider.SearchProvider;
import com.sdl.webapp.common.api.WebRequestContext;
import com.sdl.webapp.common.api.model.ViewModel;
import com.sdl.webapp.common.controller.ControllerUtils;
import com.sdl.webapp.common.controller.EntityController;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;

import javax.servlet.http.HttpServletRequest;
import java.util.Map;

import static org.apache.commons.lang3.StringUtils.isEmpty;

@Controller
@RequestMapping(ControllerUtils.INCLUDE_PATH_PREFIX + "Search/Search")
public class SearchController extends EntityController {

    @Autowired
    private SearchProvider searchProvider;

    @Autowired
    private WebRequestContext webRequestContext;

    @Override
    protected ViewModel enrichModel(ViewModel model, HttpServletRequest request) throws Exception {
        model = super.enrichModel(model, request);

        if (model == null || !(model instanceof SearchQuery)) {
            throw new DxaSearchException("Unexpected View Model. Expecting type SearchQuery but was " + model);
        }

        String queryText = request.getParameter("q");
        if (isEmpty(queryText)) {
            return model;
        }
        Map<String, String[]> queryStringParameters = request.getParameterMap();
        String start = request.getParameter("start");

        SearchQuery searchQuery = searchProvider.buildSearchQuery(model, queryText, start, queryStringParameters);
        searchProvider.executeQuery(searchQuery, webRequestContext.getLocalization());

        return searchQuery;
    }
}
